packages/flutter/lib/src/material/combo_box.dart

```diff
@@ -0,0 +1,747 @@
+
+  @override
+  void initState() {
+    super.initState();
+    _enableFilter = widget.enableFilter;
+    filteredEntries = widget.comboBoxEntries;
+    for (final ComboBoxEntry entry in filteredEntries) {
+      if (entry.enabled) {
+        _menuHasEnabledItem = true;
+        break;
+      }
+    }
+    refreshLeadingPadding();
+  }
+
+  @override
+  void didUpdateWidget(ComboBox oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.comboBoxEntries != widget.comboBoxEntries) {
+      _menuHasEnabledItem = false;
+      for (final ComboBoxEntry entry in filteredEntries) {
+        if (entry.enabled) {
+          _menuHasEnabledItem = true;
+          break;
+        }
+      }
+    }
+    if (oldWidget.leadingIcon != widget.leadingIcon) {
+      refreshLeadingPadding();
+    }
+  }
+
+  void refreshLeadingPadding() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      setState(() {
+        leadingPadding = getWidth(_leadingKey);
+      });
+    });
+  }
+
+  double? getWidth(GlobalKey key) {
+    final BuildContext? context = key.currentContext;
+    if (context != null) {
+      final RenderBox box = context.findRenderObject()! as RenderBox;
+      return box.size.width;
+    }
+    return null;
+  }
+
+  List<ComboBoxEntry> filter(List<ComboBoxEntry> entries, TextEditingController textEditingController) {
+    final List<ComboBoxEntry> filteredEntries = entries.where((ComboBoxEntry entry)
+      => entry.label.toLowerCase().contains(textEditingController.text.toLowerCase())).toList();
+
+    return filteredEntries;
+  }
+
+  int? search(List<ComboBoxEntry> entries, TextEditingController textEditingController) {
+    for (int i = 0; i < entries.length; i++) {
+      final ComboBoxEntry entry = entries[i];
+      if (textEditingController.value.text.isNotEmpty
+          && entry.label.toLowerCase().contains(textEditingController.text.toLowerCase())) {
+        return i;
+      }
+    }
+
+    return null;
+  }
+
+  List<Widget> _buildButtons(
+    List<ComboBoxEntry> filteredEntries,
+    TextEditingController textEditingController,
+    TextDirection textDirection,
+    { int? focusedIndex }
+  ) {
+    final List<Widget> result = <Widget>[];
+    final double padding = leadingPadding ?? _kDefaultHorizontalPadding;
+    final ButtonStyle defaultStyle;
+    if (textDirection == TextDirection.ltr) {
+      defaultStyle = MenuItemButton.styleFrom(
+        padding: EdgeInsets.only(left: padding, right: _kDefaultHorizontalPadding),
+      );
+    } else {
+      defaultStyle = MenuItemButton.styleFrom(
+        padding: EdgeInsets.only(left: _kDefaultHorizontalPadding, right: padding),
+      );
+    }
+
+    for (int i = 0; i < filteredEntries.length; i++) {
+      final ComboBoxEntry entry = filteredEntries[i];
+      ButtonStyle effectiveStyle = entry.style ?? defaultStyle;
+      final Color foregroundColor = effectiveStyle.foregroundColor?.resolve(<MaterialState>{MaterialState.focused})
```

Comment:
Why are are resolving the foreground color against `focused`? Shouldn't that only be done when we know the entry is focused?

Reply:
For this part, because the text field should always be focused during up/down key traversal, I just apply the "focused"  color to the background color to simulate a "focused" state for the menu item. So when we traverse the list via the up/down key, the according item can be highlighted with the focused color, and at the same time, we can use the mouse to hover the item which has a hovered overleyColor. The focus issue should be fixed once the `OverlayPortal` gets finished.

As for this line of code, if there's a custom foreground color then the "focused" color which is a background color now, should also change based on the foreground color. So I added the code above to handle this scenario. Please let me know if there are any questions:)

Reply:
Ah, I think I got it. Perhaps a better name for this might be `focusedBackgroundColor` instead to avoid confusion.